driver-core/src/main/com/mongodb/internal/async/function/AsyncCallbackSupplier.java

```diff
@@ -0,0 +1,114 @@
+/*
+ * Copyright 2008-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.mongodb.internal.async.function;
+
+import com.mongodb.annotations.NotThreadSafe;
+import com.mongodb.internal.async.SingleResultCallback;
+
+import java.util.function.Supplier;
+
+/**
+ * An {@linkplain AsyncCallbackFunction asynchronous callback-based function} of no parameters.
+ * This class is a callback-based counterpart of {@link Supplier}.
+ * Any asynchronous callback function with parameters may be represented this way by partially applying the function to its parameters
+ * until no parameters are left unapplied, and only a callback is left to be consumed.
+ *
+ * @param <R> See {@link AsyncCallbackFunction}.
+ * @see AsyncCallbackFunction
+ */
+@FunctionalInterface
+public interface AsyncCallbackSupplier<R> {
```

Comment:
Initially this class was added in the successive PR https://github.com/mongodb/mongo-java-driver/pull/784, but I needed it here to write unit tests for `LoopState`/`RetryState`. So I had to "back port" the class here, and will pull the changes to https://github.com/mongodb/mongo-java-driver/pull/784 to resolve conflicts.